File: benchmarks/correctness/real_world/r/sequence_stats.R
library(Biostrings)
library(jsonlite)
seqs <- readDNAStringSet("real_data/yeast_genome.fa")
lengths <- width(seqs)
total <- sum(lengths)
gc_weighted <- sum(sapply(1:length(seqs), function(i) {
  freq <- alphabetFrequency(seqs[[i]])
  gc <- (freq["G"] + freq["C"]) / sum(freq[c("A","C","G","T")])
  gc * lengths[i]
}))
overall_gc <- gc_weighted / total
sorted_lengths <- sort(lengths, decreasing = TRUE)
cumsum_lengths <- cumsum(sorted_lengths)
n50 <- sorted_lengths[min(which(cumsum_lengths >= total / 2))]
cat(toJSON(list(n_sequences = length(seqs), total_length = total, n50 = as.integer(n50), gc_content = overall_gc), auto_unbox = TRUE))
cat("\n")

File: benchmarks/correctness/real_world/python/sequence_stats.py
import json
from Bio import SeqIO
from Bio.SeqUtils import gc_fraction

lengths = []
gc_weighted = 0.0
for record in SeqIO.parse("real_data/yeast_genome.fa", "fasta"):
    l = len(record.seq)
    lengths.append(l)
    gc_weighted += gc_fraction(record.seq) * l

lengths.sort(reverse=True)
total = sum(lengths)
overall_gc = gc_weighted / total

# N50
cumulative = 0
n50 = 0
for l in lengths:
    cumulative += l
    if cumulative >= total / 2:
        n50 = l
        break

print(json.dumps({"n_sequences": len(lengths), "total_length": total, "n50": n50, "gc_content": overall_gc}))
